### DISCUSSION TITLE
Enum collection binding with CompiledBinding.

### DISCUSSION THREAD

**Discussion by FolderMaster** [AvaloniaUI/Avalonia#17017] [Q&A] [answered]
I have a DataTemplate, which contains a ComboBox with an Enum in SelectedItem. I would like to provide a collection of enumerations Enum.GetValues(type) in ItemsSource without creating a separate property in the interface that will store this collection, with support for CompiledBindings.
x:Static, converters do not work: when selecting an item in the ComboBox, the value is not assigned to the property.
I tried to create my own MarkupExtension with CompiledBinding, but it also does not work:
```
public class EnumsExtension : MarkupExtension
{
    private readonly Type _type;

    public EnumsExtension(Type type) => _type = type;

    public override object ProvideValue(IServiceProvider serviceProvider)
    {
        var binding = new CompiledBindingExtension()
        {
            Mode = BindingMode.OneTime,
            DataType = typeof(IEnumerable<>).MakeGenericType(_type),
            Source = Enum.GetValues(_type)
        };
        return binding;
    }
}
```
How can I solve the problem?

**Comment by rabbitism** [comment]
Ursa provides a EnumSelector control.  you can refer to its design. it is AOT compatible. 

https://github.com/irihitech/Ursa.Avalonia/tree/main/src/Ursa/Controls/EnumSelector

**Reply by FolderMaster** [reply]
It doesn't work with CompiledBinding.

**Comment by FolderMaster** [comment]
Looks more like a bug. Tried with an additional property that contained an Enum collection, but it doesn't work. It doesn't work not only with ComboBox, but also with ListBox, that is, with list controls that have a selected element.

**Comment by kyuranger** [comment]
Not sure if this is what you want, I usually initialize enum collection like this: 

```
[ObservableProperty]
ObservableCollection<NetworkTypeEnum> _AvailableNetworkTypesOC = new ObservableCollection<NetworkTypeEnum>(Enum.GetValues(typeof(NetworkTypeEnum)).Cast<NetworkTypeEnum>());
```

And then bind the `ObservableCollection` to the `ItemsSource`.

This is also available in CompiledBinding.

**Reply by FolderMaster** [reply]
Tried with ObservableCollection, didn't work. I don't know where it worked for you.
In my project, where the AvaloniaUseCompiledBindingsByDefault setting is enabled, assignment to the Enum property via Items control in DataTemplate stops working.

**Reply by IanRawley** [reply]
What exactly do you mean by "stops working"?  Could you post the XAML in question for the ComboBox?  ItemControls with selection can get very particular about the SelectedItem and ItemsSource, so seeing the XAML or how the control is setup can help diagnosing problems.  My suspicion is that the SelectedItem is being bound before the ItemsSource, causing the control to instead set SelectedItem to null because it's not present in the as-yet unbound Items.

**Reply by FolderMaster** [reply]
It seems like I bind SelectedItem to ItemsSource. I tried to do the opposite, it didn't work either.
Link to axaml: https://github.com/FolderMaster/TaskManagerApp/blob/master/View/Assets/DataTemplates.axaml
You can see the whole project.

**Reply by kyuranger** [reply]
> It seems like I bind SelectedItem to ItemsSource. I tried to do the opposite, it didn't work either. Link to axaml: https://github.com/FolderMaster/TaskManagerApp/blob/master/View/Assets/DataTemplates.axaml You can see the whole project.

I haven't tried using this with a DataTemplate, so I can't give you a proper answer. BTW, why not bind the itemsource to a ObservableCollection directly, This will make your code much simpler and clearer.

**Reply by FolderMaster** [reply]
The additional property pollutes the original class and doesn't work either.

**Reply by IanRawley** [reply, marked as answer]
Your Status property in ITask [doesn't have a Setter](https://github.com/FolderMaster/TaskManagerApp/blob/dc7a5065c83acca0a4a17bf56d97dce26ba05ba1/Model/ITask.cs#L11), that's why you can't set it.
```
        public TaskStatus Status { get; }
```

Everything else works because both Getter and Setter are specified in the more specific ITaskElement interface.  Only Status is left entirely inherited.

**Reply by FolderMaster** [reply]
It wasn't this interface that was the problem, since I use interfaces derived from it, but you were right that there was no setter for the property.
